src/internal/cluster/zarf.go

```diff
@@ -99,24 +98,129 @@ func (c *Cluster) StripZarfLabelsAndSecretsFromNamespaces() {
+	if !packageNeedsWait {
+		return types.DeployedPackage{}, nil
+	}
+
+	// Timebox the amount of time we wait for a webhook to complete before erroring
+	waitDuration := types.DefaultWebhookWaitDuration
+	if waitSeconds > 0 {
+		waitDuration = time.Duration(waitSeconds) * time.Second
+	}
+	timeout := time.After(waitDuration)
+
+	// We need to wait for this package to finish having webhooks run, create a spinner and keep checking until it's ready
+	spinner := message.NewProgressSpinner("Waiting for webhook '%s' to complete for component '%s'", hookName, component.Name)
+	defer spinner.Stop()
+	for packageNeedsWait {
+		select {
+		// On timeout, abort and return an error.
+		case <-timeout:
+			return types.DeployedPackage{}, errors.New("timed out waiting for package deployment to complete")
+		default:
+			// Wait for 1 second before checking the secret again
+			time.Sleep(1 * time.Second)
+			deployedPackage, err = c.GetDeployedPackage(deployedPackage.Name)
+			if err != nil {
+				return types.DeployedPackage{}, err
+			}
+			packageNeedsWait, _, _, err = c.PackageSecretNeedsWait(deployedPackage, component, skipWebhooks)
+			if err != nil {
+				return types.DeployedPackage{}, err
+			}
+		}
+	}
+
+	spinner.Success()
+	return deployedPackage, nil
+}
+
 // RecordPackageDeployment saves metadata about a package that has been deployed to the cluster.
-func (c *Cluster) RecordPackageDeployment(pkg types.ZarfPackage, components []types.DeployedComponent, connectStrings types.ConnectStrings) error {
-	// Generate a secret that describes the package that is being deployed
+func (c *Cluster) RecordPackageDeployment(pkg types.ZarfPackage, components []types.DeployedComponent, connectStrings types.ConnectStrings, generation int) (deployedPackage types.DeployedPackage, err error) {
 	packageName := pkg.Metadata.Name
-	deployedPackageSecret := c.GenerateSecret(ZarfNamespaceName, config.ZarfPackagePrefix+packageName, corev1.SecretTypeOpaque)
+
+	// Generate a secret that describes the package that is being deployed
+	secretName := config.ZarfPackagePrefix + packageName
+	deployedPackageSecret := c.GenerateSecret(ZarfNamespaceName, secretName, corev1.SecretTypeOpaque)
 	deployedPackageSecret.Labels[ZarfPackageInfoLabel] = packageName
 
-	stateData, _ := json.Marshal(types.DeployedPackage{
+	// Attempt to load information about webhooks for the package
+	existingPackageSecret, err := c.GetDeployedPackage(packageName)
+	if err != nil {
+		message.Debugf("Unable to fetch existing secret for package '%s': %s", packageName, err.Error())
+	}
+	componentWebhooks := existingPackageSecret.ComponentWebhooks
+
+	deployedPackage = types.DeployedPackage{
 		Name:               packageName,
 		CLIVersion:         config.CLIVersion,
 		Data:               pkg,
 		DeployedComponents: components,
 		ConnectStrings:     connectStrings,
-	})
+		Generation:         generation,
+		ComponentWebhooks:  componentWebhooks,
+	}
 
+	stateData, err := json.Marshal(deployedPackage)
```

Comment:
```suggestion
	packageData, err := json.Marshal(deployedPackage)
```
I know stateData was the old name but using a different word may be useful to disambiguate between this and zarf state.

Reply:
Updated to packageData